docs/site/migration/boot-scripts.md

```diff
@@ -6,8 +6,249 @@ sidebar: lb4_sidebar
 permalink: /doc/en/lb4/migration-boot-scripts.html
 ---
 
-{% include note.html content="
-This is a placeholder page, the task of adding content is tracked by the
-following GitHub issue:
-[loopback-next#3957](https://github.com/strongloop/loopback-next/issues/3957)
-" %}
+In LoopBack 3, predefined boot scripts are organized in the `/server/boot`
+directory, and are executed right before the server starts to perform some
+custom application initialization. The same functionality can also be achieved
+in LoopBack 4 application by adding observers.
+
+The LoopBack 3 boot script is mapped to LoopBack 4 observer's `start` function.
+And moreover, besides participating in an application's start phase, an observer
+also allows you to perform logic in the stop phase.
+
+## Migrating Boot Scripts
+
+### LoopBack 3 Boot Script
+
+LoopBack 3 has different signatures for synchronized and asynchronized boot
+scripts:
+
+They both have the first parameter as `server`(some people are used to name it
+as `app`). A synchronized script only has one parameter, for example, the
+following is a boot script that prints out a datasource's name before
+application runs:
+
+```js
+// a typical synchronized boot script
+// in file `server/boot/print-info.js`
+'use strict';
+
+module.exports = function printInfo(server) {
+  // SUPPOSE your application has a datasource called `db`, which is usually the default
+  // memory datasource in a sample LoopBack application created by CLI, regardless it's
+  // an LB3 app or LB4 app
+  const db = server.datasources.db;
+  console.log('This is a synchronized script.');
+  console.log('Your app has a datasource called: ', db.name);
+};
+```
+
+While a typical asynchronized script has a callback function as the second
+parameter:
+
+```js
+// a typical asynchronized boot script
+// in file `server/boot/create-sample.js`
+'use strict';
+
+module.exports = function createSample(server, done) {
+  // SUPPOSE your application has a Todo model
+  const Todo = server.models.Todo;
+  const sample = {title: 'a todo sample', desc: 'Something to do.'};
+  Todo.create(sample)
+    .then(result => {
+      console.log('Sample created as: ', result);
+    })
+    .then(done);
+};
+```
+
```

Comment:
I believe that boot scripts can use `async` functions too (or in fact any function that returns a promise). Can you please mention this fact and perhaps show a code snippet too?

Also in the callback-based code snippet above, please don't mix promises with callbacks, the conversion from a promise to a callback is tricky. For example, your code above does not handle errors (promise rejections) and will trigger unhandled promise rejection warning.

A better example of a callback-based boot script:

```ts
module.exports = function createSample(server, done) {
  // SUPPOSE your application has a Todo model
  const Todo = server.models.Todo;
  const sample = {title: 'a todo sample', desc: 'Something to do.'};
  Todo.create(sample, (err, result) => {
    if (err) return done(err);
    console.log('Sample created as: ', result);
    done();
  });
 };
 ```

An example of a promise-based boot script:

```ts
module.exports = function createSample(server) {
  // SUPPOSE your application has a Todo model
  const Todo = server.models.Todo;
  const sample = {title: 'a todo sample', desc: 'Something to do.'};
  return Todo.create(sample)
    .then(result => {
      console.log('Sample created as: ', result);
    });
 };
 ```

Reply:
Good point, refactored this part to be:

- sync script
- async script with cb
- async script returning a promise